### PR TITLE
[core] Fix default props behavior on all pickers

### LINKED ISSUE #4454: [pickers] DateTimePicker / TimePicker and DatePicker do not use the default props from the theme for optional props
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

If a prop has a default value set in `useXXXPickerDefaultizedPros`, then this default values takes priority over the value from the theme.

### Expected behavior 🤔

The theme value should have priority over the default value 

### Steps to reproduce 🕹

Go to [this example](https://codesandbox.io/s/date-and-time-pickers-forked-vy4u71?file=/src/App.tsx)
The view should be "year" not "day"

### Context 🔦

_No response_

### Your environment 🌎

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

### Order ID 💳 (optional)

_No response_

### LINKED ISSUE #4455: [DateTimePicker] showToolbar default value is `true`, should be `false`
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

The default value for `showToolbar` is `true`.

### Expected behavior 🤔

The default value should be `false` like the documentation suggests.

### Steps to reproduce 🕹

1. Look at the documentation of any DateTimePicker component (https://mui.com/api/desktop-date-time-picker/)
2. It says that `showToolbar` is `false` by default
3. See actual default value here: https://codesandbox.io/s/priceless-haslett-vg1bi?file=/src/Demo.tsx

### Context 🔦

This is an issue with `DateTimePicker`, `MobileDateTimePicker` and `DesktopDateTimePicker`. It seems to work as specified in documentation for `DatePicker` and `TimePicker` components.

### Your environment 🌎

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

### PR DESCRIPTION
**Pull request by flaviendelangle**
## Change `showToolbar` default value on `DateTimePicker`

Fixes #4455

It is simply a migration of https://github.com/mui/material-ui/pull/30258
The default value was `yet` but the doc was saying otherwise

image

## Apply same code structure for the `DateRangePicker` as on the other components

All others responsive picker components (`TimePicker` / `DatePicker` and `DateTimePicker`) are just tiny wrappers that calls there mobile and desktop components.

But for the range picker, we had duplication of a large interface + duplication of most code between `DateRangePicker`, `MobileDateRangePicker` and `DesktopDateRangePicker`.

## Fix the props priority (component props > theme props > default props)

Fixes #4454

We currently do something like:

```ts
const useSomePickerDefaultizedProps = (
  { someProp = defaultValue, ...other },
  name
) => {
  return useThemeProps({ 
    props: { someProp, ...other },
    name
  })
}
```

So `someProp` is never undefined when passed to `useThemeProps`.

I changed it to:

```ts
const useSomePickerDefaultizedProps = (props, name) => {
  const themeProps = useThemeProps({ props, name })

  return { someProp: defaultValue, ...themeProps }
}
```

## Others

I also replaced the ternaries by `if` in the other components.
I don't like the usage of ternaries for long expressions.

### CODE REVIEW

File: packages/x-date-pickers-pro/src/DateRangePicker/DateRangePicker.tsx [hunk 1]
```diff
@@ -108,78 +42,33 @@ export const DateRangePicker = React.forwardRef(function DateRangePicker<TDate>(
...
-    minDate: minDateProp,
-    maxDate: maxDateProp,
+    cancelText,
+    desktopModeMediaQuery = '@media (pointer: fine)',
+    DialogProps,
+    okText,
+    PopperProps,
+    showTodayButton,
+    todayText,
+    TransitionComponent,
```
Comment (alexfauquette) [inline packages/x-date-pickers-pro/src/DateRangePicker/DateRangePicker.tsx:52]: Do you know why this prop is not transmitted to the rendered components event if it's a prop that can be used with `<DesktopDateRangePicker/>`?

File: packages/x-date-pickers/src/DateTimePicker/shared.ts [hunk 2]
```diff
@@ -80,56 +79,43 @@ export function useDateTimePickerDefaultizedProps<
...
+    openTo: 'day',
+    views: ['year', 'day', 'hours', 'minutes'],
+    minDate: themeProps.minDateTime ?? defaultDates.minDate,
+    maxDate: themeProps.maxDateTime ?? defaultDates.maxDate,
+    minTime: themeProps.minDateTime,
+    maxTime: themeProps.maxDateTime,
+    ampmInClock: true,
+    showToolbar: false,
+    allowSameDateSelection: true,
+    mask: '__/__/____ __:__',
```
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply packages/x-date-pickers/src/DateTimePicker/shared.ts:112]: This one is tricky and the one it have on my PR is indeed buggy

How about this ?

```ts
    minDate: themeProps.minDateTime ?? themeProps.minDate ?? defaultDates.minDate,
    maxDate: themeProps.maxDateTime ?? themeProps.maxDate ?? defaultDates.maxDate,
    minTime: themeProps.minDateTime ?? themeProps.minTime,
    maxTime: themeProps.maxDateTime ?? themeProps.maxTime,
```

In the end we should have 

`props.minDateTime > theme.minDateTime > props.minDate > theme.minDate`
(`>` meaning "have priority over)

Same for `max`

File: packages/x-date-pickers-pro/src/DateRangePicker/DateRangePicker.tsx (hunk 1 above)
Comment (flaviendelangle) [reply packages/x-date-pickers-pro/src/DateRangePicker/DateRangePicker.tsx:52]: (added)

File: packages/x-date-pickers/src/DateTimePicker/shared.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexfauquette) [reply packages/x-date-pickers/src/DateTimePicker/shared.ts]: Yes, and it should be after the `...themeProps`

File: packages/x-date-pickers/src/DateTimePicker/shared.ts [hunk 3]
```diff
@@ -102,10 +102,10 @@ export function useDateTimePickerDefaultizedProps<
...
     views: ['year', 'day', 'hours', 'minutes'],
-    minDate: themeProps.minDateTime ?? defaultDates.minDate,
-    maxDate: themeProps.maxDateTime ?? defaultDates.maxDate,
-    minTime: themeProps.minDateTime,
-    maxTime: themeProps.maxDateTime,
+    minDate: themeProps.minDateTime ?? themeProps.minDate ?? defaultDates.minDate,
+    maxDate: themeProps.maxDateTime ?? themeProps.maxDate ?? defaultDates.maxDate,
+    minTime: themeProps.minDateTime ?? themeProps.minTime,
+    maxTime: themeProps.maxDateTime ?? themeProps.maxTime,
     ampmInClock: true,
```
Comment (alexfauquette) [inline packages/x-date-pickers/src/DateTimePicker/shared.ts:109]: Must be placed after `...themeProps`

File: packages/x-date-pickers/src/DateTimePicker/shared.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply packages/x-date-pickers/src/DateTimePicker/shared.ts]: Done

**Review by alexfauquette — APPROVED** [review]

**Pull request merged**